Adding Morphological Dilation and Erosion Algorithms

ayushbansal07 opened the pull request:
### Description
Morphological Dilation and Erosion algorithms implemented (for grayscale images).
This PR is also meant to be my solution for the competency test for GSoC 2020.

**Dilation/Erosion** process is performed by laying the structuring element H on the image I and sliding it across the image in a manner similar to convolution.

1. **Morphological Dilation** - The grayscale morphological dilation formula is written as follows :
\[𝐼⊕𝐻\](𝑢,𝑣)=max(𝑖,𝑗)∈𝐻{𝐼(𝑢−𝑖,𝑣−𝑗)+𝐻(𝑖,𝑗)}
2. **Morphological Erosion** - The grayscale morphological erosion formula is written as follows :
\[𝐼⊕𝐻\](𝑢,𝑣)=min(𝑖,𝑗)∈𝐻{𝐼(𝑢−𝑖,𝑣−𝑗)+𝐻(𝑖,𝑗)}

### References

- https://lists.boost.org/Archives/boost/2020/01/248134.php thread with initial discussion
- https://en.wikipedia.org/wiki/Dilation_(morphology)#Grayscale_dilation
- https://en.wikipedia.org/wiki/Erosion_(morphology)

### Further development during GSoC 2020

* [Documentation](https://github.com/boostorg/gil/pull/430#issuecomment-583852074)
* [KernelConcept or similar which partially satisfies the requirements of SequenceContainer](https://github.com/boostorg/gil/pull/430#discussion_r381472734)
* Final review, approval and merge of this PR

### Tasklist

<!-- Add YOUR OWN TASK(s), especially if your PR is a work in progress -->

- [x] Add basic test case(s)
- [x] Add more advanced test case(s)
- [x] Ensure all CI builds pass
- [ ] Review and approve

ayushbansal07 commented on include/boost/gil/image_processing/morph.hpp:
```diff
@@ -0,0 +1,106 @@
+#include <boost/assert.hpp>
+#include <boost/gil/image_view_factory.hpp>
+#include <boost/gil.hpp>
+#include <boost/multi_array.hpp>
```
I actually wanted to discuss this. I wanted to know your opinion on what data structure should we use to define a Structuring Element. Any binary matrix like structure would work just fine. Do you think using a simple array/vector here would be more appropriate?

stefanseefeld replied:
I agree with @mloskot 's objection, and I admit I'm a bit confused after a (rather quick, I admit) look at this PR: you include the `boost/multi_array.hpp` header, but you never use it (explicitly).
Rather, you use a template argument `SE`, which in itself sounds perfectly fine, but I think requires to specify exactly what kind of requirements you put onto that type.
Could we get away with using the same requirements we put on `SrcView` ?

ayushbansal07 replied:
Ohh, yes it was a mistake to have included `boost/multi_array.hpp` in current code. I was actually playing around with multi_array to figure out a way to put up a requirement check on it and I forgot to remove that header in the end. However, in the next iteration I would be requiring it as i have included this check - 
`boost::function_requires<boost::multi_array_concepts::ConstMultiArrayConcept<StructElement, 2>>();`
@mloskot  @stefanseefeld could you please give your comments on whether using multi_array as the type of SE and using this check is sufficient and correct?

stefanseefeld replied:
Well, as I said, I think requiring multi_array would be wrong. The whole point of parametrizing these functions with `SE` is to let users fill in that parameter with their own type, so long as it meets the requirements.
(For avoidance of doubt: using multi_array in the example is perfectly fine.)

mloskot replied:
@ayushbansal07
> could you please give your comments on whether using multi_array as the type of SE and using this check is sufficient and correct?

As I poined earlier, I also agree with @stefanseefeld on this: please try to avoid `multi_array` in codebase of GIL core and tests, and try to make those possibly container-agnostic. In examples, it is fine.

ayushbansal07 replied:
@mloskot @stefanseefeld I understand your point now. I have updated morphology.hpp. I now model a SE to be of type vector or any similar structure (which can support .size() functionality).

mloskot replied:
@ayushbansal07 
> I now model a SE to be of type vector or any similar structure (which can support .size() functionality).

As per one of the assertions, there is also required subscript (index) operator on element

ayushbansal07 replied:
@mloskot 
I am unable to find any check for subscript operator. I however did find a check for Container type.I wondered if checking for a Container type. So would using that and replacing all my occurrences of `elem[i]` with `elem.begin()+i` be a good thing to do?

mloskot replied:
Ideally if there is a concept representing requirements of `StructElement`, but I wouldn't bother for now.

For now, I'd simply add (Doxygen-style) comments to `erode` and `dilate` explaining the requirement, similar to https://github.com/boostorg/gil/blob/8235ee645a4e8c6e20722cc9c9eebe1f8050f9b6/include/boost/gil/extension/numeric/kernel.hpp#L32

In future, structuring element could be defined similarly to the kernel classes.
Since your algorithm now sort of assumes rectangular shape of the structured element, it could reuse the `kernel_2d`, but I don't think it is necessary or even a good idea considering future improvements (e.g. ovals and other non-rectangular shape of structuring elements 😄).

mloskot replied:
FYI, I'm marking this conversation as resolved

ayushbansal07 commented:
Updates - 
1. Resolved all the trivial coding issues.
2. Added a test for erosion and Dilation. (Will add more tests in the meanwhile, marked as TODO).
3. Removed .png files.
Also, I modified morphology.hpp to better handle the corner cases. 

TODOs - 
- [ ] Add more tests.

mloskot commented:
@ayushbansal07 
> Updates -
>  ...

This is looking good. Thanks for your efforts to improve this. I'll get another look and review near Wednesday.

> TODOs -
> ...

I've taken the liberty and added those as the PR tasklist which you can also edit.

------

Another thing, you mentioned on the Boost mailing list you will be proposing those algorithms as part of your **competency test**. If you are working on this PR as the competency test, please update the description to make it clear this is your solution for the GSoC 2020 competency test.

Since working on PR may involve _much more effort_ than working just on C++ code of the competency test solution, you are not expected to finish-polish the PR in order to count it for a completed competency test solution. In other words, your PR does not have to be approved and merged to consider the test completed. For the competency test, the gist of C++ code is most important.

Although you can postpone addressing further reviews and completion of this PR for later (e.g. during GSoC), you are very welcome to continue working on to make it ready for merging. It's your call.

/cc @stefanseefeld unless he does not agree and would like to clear my explanation.

ayushbansal07 commented:
@mloskot 
> please update the description to make it clear this is your solution for the GSoC 2020 competency test.

I have updated the description. 

I'd be happy to continue working on this and want to try my best to get it to merge before submitting my GSoC Proposal :)

ayushbansal07 commented on example/morphology.cpp:
```diff
@@ -0,0 +1,36 @@
+#include <boost/gil.hpp>
+#include <boost/multi_array.hpp>
+
+#define SE_SIZE 3
```
Sorry, college habbits :P

ayushbansal07 commented:
@mloskot 
I have fixed the mentioned nitpicks.

> Here you can see comparison of various pixel iteration methods available in GIL
Thank you for this interesting comparison. If you don't mind me asking, I wanted to know, just of out curiosity, what could be the reason behind such huge run time of 2d_y_iterator. Is it because the each iteration would result in a cache miss?

mloskot commented:
@ayushbansal07 
> what could be the reason behind such huge run time of 2d_y_iterator.
> Is it because each iteration would result in a cache miss?

Yes. I recommend to watch the original video lecture about GIL presented by Lubomir Bourdev
which is hosted on YouTube and linked from the documentation here https://boostorg.github.io/gil/html/tutorial/video.html
The Y-iterator and cache issue starts from around 8:00.

ayushbansal07 commented:
@mloskot 
I have added advanced test cases too. Please let me know further steps to complete this PR :-)

ayushbansal07 commented:
Would it be okay if I did the documentation part later (during GSoC time)? I was asking this because I was planning to implement a separate class for Structuring Element as first step of my GSOC project and some of those corner cases, etc. would depend on the design of Structuring Element class. So, I was wondering if all of that could be done together afterwards.

mloskot commented:
> Would it be okay if I did the documentation part later (during GSoC time)? 
> (...)
> So, I was wondering if all of that could be done together afterwards.

Yes, certainly. That's what I also suggested in https://github.com/boostorg/gil/pull/430#issuecomment-581576130

I think, your PR is completed from the competency test stand point, but to get it merged it will need that extra bit of work during the GSoC. So, let's keep it open as work in progress for now.

/cc @stefanseefeld

mloskot commented on example/morphology.cpp:
```diff
@@ -0,0 +1,37 @@
+#include <boost/gil/extension/io/png.hpp>
```
Comment with licence and your copyright is missing

mloskot commented on include/boost/gil/image_processing/morphology.hpp:
```diff
@@ -0,0 +1,131 @@
+#define BOOST_GIL_IMAGE_PROCESSING_MORPHOLOGY_HPP
+
+#include <boost/assert.hpp>
+#include <boost/gil.hpp>
```
We should avoid including the all-in-one header `#include <boost/gil.hpp>`
If this PR is accepted, it will include `morphology.hpp` leading to cyclic dependency.
Could you try to include individual headers of stuff that is used (yes, it may be tricky)?

ayushbansal07 replied:
@mloskot I had a little doubt about this. After I removed `#include <boost/gil.hpp>`, my example and test files compiled without any errors. I was under the impression that definitions for concepts such as `ImageViewConcept` would be included from the all-in-one header, but I guess that is not the case. There should have been an error according to me, but wasn't. Is there something that I am missing?

mloskot replied:
@ayushbansal07 You are correct to be suspicious.

I have just fixed one issue in `Jamfile`, https://github.com/boostorg/gil/commit/d026d50e178eecf6d2468f955ccaf4983ada9663, so please update your branch to get this fix.

Now, if you build the standalone headers tests - tests that verify each header can be compiled alone - `b2 cxxstd=11 test/core/headers`, then you should see that your header does not compile:

![image](https://user-images.githubusercontent.com/80741/75064367-17b2e500-54e7-11ea-85be-049588dd666e.png)

I also tried to compile your `example/morphology.cpp` using MSVC from VS2019 and it is also failing:

![image](https://user-images.githubusercontent.com/80741/75064655-a3c50c80-54e7-11ea-8f42-768b6460b007.png)

ayushbansal07 replied:
I have now fixed all the errors and warnings for both example and test file.

mloskot replied:
Thanks

mloskot commented on include/boost/gil/image_processing/morphology.hpp:
```diff
@@ -0,0 +1,131 @@
+
+// SE refers to Structuring Element
+// Result type after convolution with the structuring element
+enum class morphology_operator{
```
Nitpick:

```
enum class morphology_operator{
```
to
```
enum class morphology_operator
{
```

mloskot commented on include/boost/gil/image_processing/morphology.hpp:
```diff
@@ -0,0 +1,131 @@
+        // (src_y + struct_elem_y - struct_elem_height/2, src_x + struct_elem_x - struct_elem_width/2)
+        // NOTE: This calculation is only valid when the struct_elem dimensions are odd.
+        auto y_to_read = src_y + struct_elem_y - struct_elem_height/2;
+        if(y_to_read < 0 || y_to_read >= src_view.height())
```
Nitpick: `for(` -> `for (` and `if(` -> `if (`, here and everywhere.

mloskot commented on include/boost/gil/image_processing/morphology.hpp:
```diff
@@ -0,0 +1,131 @@
+        for(std::ptrdiff_t struct_elem_x = 0; struct_elem_x < struct_elem_width; ++struct_elem_x)
+        {
+            // We need to consider only those positions which have 1 in Structuring Element
+            if(struct_elem[struct_elem_x][struct_elem_y] == 1)
```
Loose comment, a while ago I personally started appreciating the early exits and continuations as recommended by https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code, this also helps to avoid too deep indentations what is a good idea for long names of templated types, etc.

ayushbansal07 replied:
I too prefer early exits and continuations. But, for this case I this way seemed more appropriate to me. The only reason being this flow is more coherent with the natural flow of the algorithm and for a new reader it would be more intuitive. And aldo, there is not much deep indentation in this part of code. However, if you still feel if I should use early continuation here, I can do that?

mloskot replied:
It's fine as it is. The number of nested levels is not large to make it a problem. I pointed at the issues to be aware as a tool to avoid code drift towards the right too much :)

mloskot commented on include/boost/gil/image_processing/morphology.hpp:
```diff
@@ -0,0 +1,131 @@
+                if (src_point_val < min)
+                    min = src_point_val;
+                if (src_point_val > max)
+                    max = src_point_val;
```
I'm curious, why these are `point`-s? Just `src_value` would be clear? Values of image are pixels...

mloskot commented on include/boost/gil/image_processing/morphology.hpp:
```diff
@@ -0,0 +1,131 @@
+/// \param src_view - The source image view.
+/// \param dst_view - The destination image view (after performing dilation).
+/// \param struct_elem - The Structuing Element. Should be a 2D-array type consiting of binary int values (0 and 1).
+/// \tparam SrcView type of source image. Should confirm to gil::ImgaeViewConcept.
```
> Should confirm to gil::ImgaeViewConcept.

should just read

> Models gil::ImgaeViewConcept.

If you are not familiar with notion of model w.r.t. to concept, here you can find nicely documented examples
https://www.boost.org/libs/geometry/doc/html/geometry/reference/concepts.html

(we need similar documentation for GIL :))

mloskot commented on include/boost/gil/image_processing/morphology.hpp:
```diff
@@ -0,0 +1,131 @@
+/// \param struct_elem - The Structuing Element. Should be a 2D-array type consiting of binary int values (0 and 1).
+/// \tparam SrcView type of source image. Should confirm to gil::ImgaeViewConcept.
+/// \tparam DstView type of destination image. Should confirm to gil::MutableImageViewConcept.
+/// \tparam StructElement type of structuring element. Should provide size() and operator[].
```
In future, for such kernel-like containers, we may have `KernelConcept` or similar which partially satisfies the requirements of [SequenceContainer](https://en.cppreference.com/w/cpp/named_req/SequenceContainer).

ayushbansal07 replied:
Will implement it during the GSoC period :)

mloskot replied:
Sure. I added "Further development during GSoC 2020" section to the PR description.

mloskot commented on test/core/image_processing/morphology.cpp:
```diff
@@ -0,0 +1,340 @@
+#include <boost/gil/image_processing/morphology.hpp>
```
Comment with licence and your copyright is missing

mloskot commented on test/core/image_processing/morphology.cpp:
```diff
@@ -0,0 +1,340 @@
+    test_dilation_big();
+    test_erosion_big();
+
+    return boost::report_errors();
```
This is a good bunch of test cases.

In future, I'd suggest to convert it to Boost.Test.

ayushbansal07 replied:
Sure. Will do it during GSOC :)

ayushbansal07 commented:
@mloskot Really sorry for the late response. I was a little busy with my semester exams. I have made the required changes :)

mloskot commented:
@ayushbansal07 
> Really sorry for the late response.

There is no need to be sorry. It's your PR, so you work on it at your own pace.

> I have made the required changes :)

Please, hit the `Resolve conversation` button for each item you address or add comment explaining a thing is not addressed. This helps to keep completed review iterations cleared.

ayushbansal07 commented:
@mloskot I have made relevant changes to make sure that the Cl builds and has minimal warnings. 

Also, FYI, I have checked **Add more advanced test case(s)** and **Ensure all CI builds pass** in the task list.

mloskot commented:
@ayushbansal07 in https://github.com/boostorg/gil/pull/430#issuecomment-581594149
> I'd be happy to continue working on this and want to try my best to get it to merge before submitting my GSoC Proposal

Thank you for the very good job you've done. I consider it as more than complete for your competency test.

Although the PR is very close to be approved and merged, we are not in rush
and I'd prefer to defer merging. I also mentioned it in https://github.com/boostorg/gil/pull/430#pullrequestreview-353384484 

My reasoning is to :
- give myself a bit more time to try your code out in action (I have only read-only reviewed it).
- give other active/interested developers and users a chance to have a look and comment (/cc @stefanseefeld, @lpranam, @simmplecoder, @miralshah365, @chhenning).
- leave it for you, in case you are approved as student, as an easy first task to completed as part of your GSoC work (an early success is an important element of GSoC :-))

Regardless if you will participate during GSoC or not, since this PR is pretty much completed, I'm fairly sure your contribution will be merged eventually.

ayushbansal07 commented:
@mloskot 
Thank you for your reviews. I too agree to defer the merging mostly because of the last reason that you mentioned :)

Also, if it is not much of a trouble for you, feel free to mention me in any discussion that might potentially help me understand more about GIL.

Thank you again! :)

mloskot commented:
@ayushbansal07 
> please feel free to mention me in any discussion that might potentially help me understand more about GIL

Sure, but nothing immediately comes to mind. Feel free to browse the issues, check the [good first issue](https://github.com/boostorg/gil/issues?q=is%3Aissue+is%3Aopen+label%3Agood-first-issue) list.

For understanding GIL, I highly recommend getting familiar with these topics from [GIL docs](https://boostorg.github.io/gil/develop/doc/html/index.html):

- The video lecture
- The three tutorials
- These from the design guide:
    - Channel
    - Color Space and Layout
    - Color Base
    - Pixel
    - Pixel Iterator
    - Pixel Locator
    - Image View
    - Image

These are essential parts to understand to be comfortable with using GIL.
If anything needs clarifying, don't hesitate to ask questions on the boost-gil list.
If you think anything is missing or wrong, please, open new issue.

This should make a busy GSoC preparation for next days :)

mloskot commented:
This is being replaced with more feature-complete morphological operations implementation in #541

Thanks @ayushbansal07 for your work here.
